<a href="https://colab.research.google.com/github/mbuitragoc/AlgorithmsUN2024I/blob/main/Lab4/mbuitragoc_Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

4 7
11


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
%%writefile output1.txt
90

Writing output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Writing input2.txt


In [ ]:
%%writefile output2.txt
90

Writing output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



To make it easier to generate the test cases we produced random tests for this

In [ ]:
import random
number = random.randint(2, 20)
print(number)
numbers = [str(random.randint(1, 200000)) for _ in range(number)]
print(' '.join(numbers))

17
60188 59331 79393 162530 36444 81978 154516 71682 57356 1788 118850 14026 145878 67212 180538 154550 132119


####Inputs for Test 4

In [ ]:
%%writefile input4.txt
5
100000 6500 23456 123456 98

Writing input4.txt


In [ ]:
%%writefile output4.txt
12345600000

Writing output4.txt


In [ ]:
!./max_pairwise_product_long_full < input4.txt > ctest4.txt
!grep -xvFf output4.txt ctest4.txt

####Inputs for Test 5

In [ ]:
%%writefile input5.txt
10
100000 200000 199999 198999 100 20 19999 19899

Writing input5.txt


In [ ]:
%%writefile output5.txt
39999800000

Writing output5.txt


In [ ]:
!./max_pairwise_product_long_full < input5.txt > ctest5.txt
!grep -xvFf output5.txt ctest5.txt

####Inputs for Test 6

In [ ]:
%%writefile input6.txt
14
137085 418 65440 153266 35820 511 135544 107425 21528 109508 137551 15103 51818 58457

Writing input6.txt


In [ ]:
%%writefile output6.txt
21081891566

Writing output6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt > ctest6.txt
!grep -xvFf output6.txt ctest6.txt

####Inputs for Test 7

In [ ]:
%%writefile input7.txt
5
27707 89651 173127 24779 197012

Writing input7.txt


In [ ]:
%%writefile output7.txt
34108096524

Writing output7.txt


In [ ]:
!./max_pairwise_product_long_full < input7.txt > ctest7.txt
!grep -xvFf output7.txt ctest7.txt

####Inputs for Test 8

In [ ]:
%%writefile input8.txt
7
181169 186907 26822 180751 140223 46947 71938

Writing input8.txt


In [ ]:
%%writefile output8.txt
33861754283

Writing output8.txt


In [ ]:
!./max_pairwise_product_long_full < input8.txt > ctest8.txt
!grep -xvFf output8.txt ctest8.txt

####Inputs for Test 9

In [ ]:
%%writefile input9.txt
14
17472 108927 14916 122218 80916 44068 172256 186078 58993 101650 54594 122494 171557 24454

Writing input9.txt


In [ ]:
%%writefile output9.txt
32053051968

Writing output9.txt


In [ ]:
!./max_pairwise_product_long_full < input9.txt > ctest9.txt
!grep -xvFf output9.txt ctest9.txt

####Inputs For Test 10


In [ ]:
%%writefile input10.txt
17
60188 59331 79393 162530 36444 81978 154516 71682 57356 1788 118850 14026 145878 67212 180538 154550 132119

Writing input10.txt


In [ ]:
%%writefile output10.txt
29342841140

Writing output10.txt


In [ ]:
!./max_pairwise_product_long_full < input10.txt > ctest10.txt
!grep -xvFf output10.txt ctest10.txt

#### Verifying the results

In [ ]:
%%writefile run_Last_tests.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {3..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_Last_tests.sh


In [ ]:
!chmod u+x run_Last_tests.sh
!./run_Last_tests.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m54.328s
user	4m51.409s
sys	0m0.117s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m54.203s
user	4m51.342s
sys	0m0.143s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177091


In [ ]:
# Estimated time by instruction

295/200003177091


1.474976569326082e-09

In [ ]:
# Estimated number of instructions by second

200003177091/295

677976871.4949152

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.845s
user	0m0.827s
sys	0m0.002s


## Python

In [ ]:
# Here we write the max_pairwise_product_long_full in Python
def max_pairwise_product(numbers):
    max_product = 0
    aux_product = 0
    n = len(numbers)
    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

In [ ]:
with open('input1.txt', 'r') as f:
      n = int(f.readline().strip())
      numbers = list(map(int, f.readline().split()))

m = max_pairwise_product(numbers)

with open('ptest1.txt', 'w') as f:
    f.write(str(m) + '\n')

In [ ]:
# Verify
def run_tests(function):
    num_tests = 10

    for i in range(1, num_tests + 1):
        input_file = f'input{i}.txt'
        expected_output_file = f'output{i}.txt'

        with open(input_file, 'r') as f:
            n = int(f.readline().strip())
            numbers = list(map(int, f.readline().split()))

        # Calculate max pairwise product
        result = function(numbers)

        # Read expected output
        with open(expected_output_file, 'r') as f:
            expected_result = int(f.readline().strip())

        # Compare results using assert
        assert result == expected_result, f'Test {i} failed: Expected {expected_result}, got {result}'

        print(f'Test {i} passed')

In [ ]:
run_tests(max_pairwise_product)

Test 1 passed
Test 2 passed
Test 3 passed
Test 4 passed
Test 5 passed
Test 6 passed
Test 7 passed
Test 8 passed
Test 9 passed
Test 10 passed


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product< input12.txt

^C


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long< input12.txt

/bin/bash: line 1: input13.txt: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


In [ ]:
def max_pairwise_product_v2(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

In [ ]:
run_tests(max_pairwise_product_v2)

Test 1 passed
Test 2 passed
Test 3 passed
Test 4 passed
Test 5 passed
Test 6 passed
Test 7 passed
Test 8 passed
Test 9 passed
Test 10 passed


In [ ]:
def max_pairwise_product_v2_long(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

In [ ]:
run_tests(max_pairwise_product_v2_long)

Test 1 passed
Test 2 passed
Test 3 passed
Test 4 passed
Test 5 passed
Test 6 passed
Test 7 passed
Test 8 passed
Test 9 passed
Test 10 passed


### Estimate and verify the maximun n that can be solved in 1 second



11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound and we knwo we can do aproximately 677976871.4949152 instructions per second, so we have to solve the ecuation. Lo que nos deja que en el limite inferior y en el mejor de los casos podemos procesar un n de 18410, mientras que en nuestro limite superior es de solo 16467

In [ ]:
# n = 18410
n = 16467
print( "[", 11 + 7*n + 2*(n-1)*n,  ",", 11 + 7*n  + 2.5*(n-1)*n, ")")

[ 542406524 , 677979335.0 )


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [ ]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    int n = numbers.size();

    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    long long max_product = static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_sort.cpp


In [ ]:
%%writefile sort_solution_nlogn.py

n = int(input())  # 1
numbers = list(map(int, input().split()))  # n

numbers.sort()  # n*log(n)
product = numbers[-1] * numbers[-2]  # 1

print(f"Product: {product}")  # 1

Test with inputs 1 and 2



In [ ]:
%%writefile run_Last_tests.sh
#!/bin/bash
g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
for i in {3..10}
do
    ./max_pairwise_product_sort < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_Last_tests.sh


In [ ]:
!chmod u+x run_Last_tests.sh
!./run_Last_tests.sh

In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt

Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_sort< input13.txt

/bin/bash: line 1: input13.txt: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


In [ ]:
!time python sort_solution_nlogn.py < input3.txt

20 stress tests

In [ ]:
%%writefile max_pairwise_product_sort_stress.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    int n = numbers.size();

    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    long long max_product = static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];

    return max_product;
}

long long MaxPairwiseProductSlow(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			      aux_product =  (long long) numbers[first] * numbers[second];
            if (max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

std::vector<int> GenerateRandomVector(int length) {
    std::vector<int> random_vector(length);
    for (int i = 0; i < length; ++i) {
        random_vector[i] = rand() % 200001; // Random number between 0 and 200,000
    }
    return random_vector;
}

int main() {
    srand(time(0)); // Seed for random number generation

    const int rounds = 20;

    for (int round = 1; round <= rounds; ++round) {
        int vector_length = rand() % 199999 + 2; // Random vector length between 2 and 200,000

        // Generate random vectors
        std::vector<int> random_vector = GenerateRandomVector(vector_length);

        // Calculate results using both functions
        long long result_optimized = MaxPairwiseProduct(random_vector);
        long long result_slow = MaxPairwiseProductSlow(random_vector);

        // Compare results
        if (result_optimized != result_slow) {
            std::cerr << "Test failed for round " << round << "\n";
            std::cerr << "Optimized result: " << result_optimized << "\n";
            std::cerr << "Slow result: " << result_slow << "\n";
            return 1; // Exit with an error code
        }
    }

    std::cout << "All tests passed successfully!\n";

    return 0;
}

Overwriting max_pairwise_product_sort_stress.cpp


In [ ]:
!g++ max_pairwise_product_sort_stress.cpp -o max_pairwise_product_sort_stress
!./max_pairwise_product_sort_stress

All tests passed successfully!


In [ ]:
!python sort_solution_nlogn.py < sinput1.txt
!python sort_solution_nlogn.py < sinput2.txt
!python sort_solution_nlogn.py < sinput3.txt
!python sort_solution_nlogn.py < sinput4.txt
!python sort_solution_nlogn.py < sinput5.txt
!python sort_solution_nlogn.py < sinput6.txt
!python sort_solution_nlogn.py < sinput7.txt
!python sort_solution_nlogn.py < sinput8.txt
!python sort_solution_nlogn.py < sinput9.txt
!python sort_solution_nlogn.py < sinput10.txt
!python sort_solution_nlogn.py < sinput11.txt
!python sort_solution_nlogn.py < sinput12.txt
!python sort_solution_nlogn.py < sinput13.txt
!python sort_solution_nlogn.py < sinput14.txt
!python sort_solution_nlogn.py < sinput15.txt
!python sort_solution_nlogn.py < sinput16.txt
!python sort_solution_nlogn.py < sinput17.txt
!python sort_solution_nlogn.py < sinput18.txt
!python sort_solution_nlogn.py < sinput19.txt
!python sort_solution_nlogn.py < sinput20.txt

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

long long MaxPairwiseProduct(const std::vector<int>& numbers, int& count) {                               //INSTRUCTIONS EXECUTED BY LINE
    long long max_product = 0LL;                                                                          //1
    int n = numbers.size();                                                                               //1
    std::vector<int> sorted_numbers = numbers;                                                            //n
    std::sort(sorted_numbers.begin(), sorted_numbers.end());                                              //n logn
    long long largest1 = static_cast<long long>(sorted_numbers[n - 1]);                                   //2
    long long largest2 = static_cast<long long>(sorted_numbers[n - 2]);                                   //2
    max_product = largest1 * largest2;                                                                    //1

    count += 7 + n + n * log2(n);

    return max_product;                                                                                   //7 + n + n log n
}

int main() {                                                      //INSTRUCTIONS EXECUTED BY LINE
    int n;                                                        //1
    std::cin >> n;                                                //1
    std::vector<int> numbers(n);                                  //n
    for (int i = 0; i < n; ++i) {                                 //2 + 2n
        std::cin >> numbers[i];                                   //n
    }

    int count = 0;

    std::cout << MaxPairwiseProduct(numbers, count) << "\n";      //7 + n + n log n

    count += 4  + 4 * n;

    std::cout << "Total instructions executed: " << count << "\n";

    return 0;                                                     //11 + 5n + n logn

}

Overwriting max_pairwise_product_sort_long.cpp


Al ejecutar con esta version optimizada deberiamos de evidenciar 11 + 5n + n logn instrucciones


For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 5*(10) + (10) log (19)  = 94.219 $$

In [ ]:
import math
n = 10
11 + 5*(n) + 10*math.log2(n)

94.21928094887363

In [ ]:
n1 = 10
print(11 + 5*(n1) + 10*math.log2(n1))

n2 = 10
print(11 + 5*(n2) + 10*math.log2(n2))

n3 = 6
print(11 + 5*(n3) + 10*math.log2(n3))

94.21928094887363
94.21928094887363
66.84962500721156


In [ ]:
!g++ max_pairwise_product_sort_long.cpp -o max_pairwise_product_sort_long
!./max_pairwise_product_sort_long< input1.txt
!./max_pairwise_product_sort_long< input2.txt
!./max_pairwise_product_sort_long< input3.txt

90
Total instructions executed: 94
90
Total instructions executed: 94
40000000000
Total instructions executed: 56


In [ ]:
%%writefile sort_solution_nlogn.py
import math as m

instructions = 0
n = int(input())  # 1
instructions += 1
numbers = list(map(int, input().split()))  # n
instructions += 4 * n

numbers.sort()  # n*log(n)
instructions += m.log2(n)
product = numbers[-1] * numbers[-2]  # 1
instructions += 2

print(f"Product: {product}")  # 1
instructions += 1

print("Total instructions executed:", int(instructions))

In [ ]:
!python sort_solution_nlogn.py < input1.txt
!python sort_solution_nlogn.py < input2.txt
!python sort_solution_nlogn.py < input3.txt

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!time ./max_pairwise_product_sort< input1.txt
!time ./max_pairwise_product_sort< input2.txt
!time ./max_pairwise_product_sort< input3.txt

90

real	0m0.003s
user	0m0.002s
sys	0m0.001s
90

real	0m0.004s
user	0m0.001s
sys	0m0.002s
40000000000

real	0m0.004s
user	0m0.002s
sys	0m0.001s


In [ ]:
# Estimated time by instruction
0.004/94

4.2553191489361704e-05

In [ ]:
# Estimated number of instructions by second

94/0.004

23500.0

In [ ]:
%%time
!python sort_solution_nlogn.py < input1.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00477 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python sort_solution_nlogn.py < input2.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00527 / 47
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python sort_solution_nlogn.py < input3.txt

In [ ]:
instructionTime = 0.00825 / 30
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

### Estimate and verify the maximun n that can be solved in 1 second

11 + 5n + n logn and we knwo we can do aproximately 23500 instructions per second, so we have to solve the ecuation. Lo que nos deja que en valor aproximado de 16281,334

## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_efficient.cpp
#include <iostream>
#include <vector>
#include <cmath>

long long MaxPairwiseProductEfficient(const std::vector<int>& numbers, int& count) {    //INSTRUCTIONS EXECUTED BY LINE
    long long max_product = 0LL;                                                        //1
    int n = numbers.size();                                                             //1
    long long largest1 = static_cast<long long>(numbers[0]);                            //1
    long long largest2 = static_cast<long long>(numbers[1]);                            //1

    for (int i = 2; i < n; ++i) {                                                       //n - 2
        long long current_number = static_cast<long long>(numbers[i]);                  //n - 2

                                                                                        // n - 2 -> entre ambas asignaciones [0,  n-2)
        if (current_number > largest1) {                                                //1
            largest2 = largest1;
            largest1 = current_number;
            count += 3;
        } else if (current_number > largest2) {                                         //1
            largest2 = current_number;
            count += 2;
        }
    }

    max_product = largest1 * largest2;                                                  //1

    count += 5 + 3 * (n - 2); // with a +- (n-2) for the case when [0] and [1] are the biggest and we dont have to reassign

    return max_product;
}

int main() {                                                      //INSTRUCTIONS EXECUTED BY LINE
    int n;                                                        //1
    std::cin >> n;                                                //1
    std::vector<int> numbers(n);                                  //n
    for (int i = 0; i < n; ++i) {                                 //2 + 2n
        std::cin >> numbers[i];                                   //n
    }

    int count = 0;

    std::cout << MaxPairwiseProductEfficient(numbers, count) << "\n";

    count += 4 + 4 * n;

    std::cout << "Total instructions executed: " << count << "\n";

    return 0;
}

Overwriting max_pairwise_product_efficient.cpp


In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    largest = float('-inf')
    second_largest = float('-inf')

    for num in numbers:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest:
            second_largest = num

    return largest, second_largest

def main():
    n = int(input())
    numbers = list(map(int, input().split()))

    result = largest_numbers(numbers)

    print("Product:", result[0] * result[1])

if __name__ == "__main__":
    main()

Test with inputs 1 and 2



In [ ]:
%%writefile run_Last_tests.sh
#!/bin/bash
g++ max_pairwise_product_efficient.cpp -o max_pairwise_product_efficient
for i in {1..3}
do
    ./max_pairwise_product_efficient < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_Last_tests.sh


In [ ]:
!chmod u+x run_Last_tests.sh
!./run_Last_tests.sh

Total instructions executed: 97
Total instructions executed: 73
Total instructions executed: 52


In [ ]:
! python sort_efficient_solution_python.py < input1.txt
! python sort_efficient_solution_python.py < input2.txt

Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_efficient < input3.txt

40000000000
Total instructions executed: 52

real	0m0.004s
user	0m0.001s
sys	0m0.002s


In [ ]:
!time python sort_efficient_solution_python.py < input3.txt

20 stress tests

In [ ]:
%%writefile max_pairwise_product_efficient_stress.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();
    long long largest1 = static_cast<long long>(numbers[0]);
    long long largest2 = static_cast<long long>(numbers[1]);

    for (int i = 2; i < n; ++i) {
        long long current_number = static_cast<long long>(numbers[i]);
        if (current_number > largest1) {
            largest2 = largest1;
            largest1 = current_number;
        } else if (current_number > largest2) {
            largest2 = current_number;
        }
    }

    max_product = largest1 * largest2;
    return max_product;
}

long long MaxPairwiseProductSlow(const std::vector<int>& numbers) {
    int n = numbers.size();

    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    long long max_product = static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];

    return max_product;
}

std::vector<int> GenerateRandomVector(int length) {
    std::vector<int> random_vector(length);
    for (int i = 0; i < length; ++i) {
        random_vector[i] = rand() % 200001; // Random number between 0 and 200,000
    }
    return random_vector;
}

int main() {
    srand(time(0)); // Seed for random number generation

    const int rounds = 20;

    for (int round = 1; round <= rounds; ++round) {
        int vector_length = rand() % 199999 + 2; // Random vector length between 2 and 200,000

        // Generate random vectors
        std::vector<int> random_vector = GenerateRandomVector(vector_length);

        // Calculate results using both functions
        long long result_optimized = MaxPairwiseProduct(random_vector);
        long long result_slow = MaxPairwiseProductSlow(random_vector);

        // Compare results
        if (result_optimized != result_slow) {
            std::cerr << "Test failed for round " << round << "\n";
            std::cerr << "Optimized result: " << result_optimized << "\n";
            std::cerr << "Slow result: " << result_slow << "\n";
            return 1; // Exit with an error code
        }
    }

    std::cout << "All tests passed successfully!\n";

    return 0;
}

Writing max_pairwise_product_efficient_stress.cpp


In [ ]:
!g++ max_pairwise_product_efficient_stress.cpp -o max_pairwise_product_efficient_stress
!./max_pairwise_product_efficient_stress

All tests passed successfully!


In [ ]:
!python sort_efficient_solution_python.py < sinput1.txt
!python sort_efficient_solution_python.py < sinput2.txt
!python sort_efficient_solution_python.py < sinput3.txt
!python sort_efficient_solution_python.py < sinput4.txt
!python sort_efficient_solution_python.py < sinput5.txt
!python sort_efficient_solution_python.py < sinput6.txt
!python sort_efficient_solution_python.py < sinput7.txt
!python sort_efficient_solution_python.py < sinput8.txt
!python sort_efficient_solution_python.py < sinput9.txt
!python sort_efficient_solution_python.py < sinput10.txt
!python sort_efficient_solution_python.py < sinput11.txt
!python sort_efficient_solution_python.py < sinput12.txt
!python sort_efficient_solution_python.py < sinput13.txt
!python sort_efficient_solution_python.py < sinput14.txt
!python sort_efficient_solution_python.py < sinput15.txt
!python sort_efficient_solution_python.py < sinput16.txt
!python sort_efficient_solution_python.py < sinput17.txt
!python sort_efficient_solution_python.py < sinput18.txt
!python sort_efficient_solution_python.py < sinput19.txt
!python sort_efficient_solution_python.py < sinput20.txt

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile sort_efficient_solution_python.py

def largest_numbers(numbers):
    instructions = 0
    largest = float('-inf')  # 1
    instructions += 1
    second_largest = float('-inf')  # 1
    instructions += 1

    for num in numbers:  # 1 + 2n
        if num > largest:  # n
            second_largest = largest  # [0, n]
            largest = num  # [0, n]
            instructions += 2
        elif num > second_largest:  # n
            second_largest = num  # [0, n]
            instructions += 1

    return largest, second_largest, instructions

def main():
    instructions = 0

    n = int(input())  # 1
    instructions += 2
    numbers = list(map(int, input().split()))  # n
    instructions += 4*n

    result = largest_numbers(numbers)  # 1
    instructions += result[2] + 1

    print("Product:", result[0] * result[1])  # 1
    print("Total instructions executed:", instructions)

if __name__ == "__main__":
    main()

In [ ]:
!python sort_efficient_solution_python.py < input1.txt
!python sort_efficient_solution_python.py < input2.txt
!python sort_efficient_solution_python.py < input3.txt

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!python sort_efficient_solution_python.py < input1.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00479 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python sort_efficient_solution_python.py < input2.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00549 / 48
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python sort_efficient_solution_python.py < input3.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00584 / 37
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
!time python sort_efficient_solution_python.py < input12.txt

## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [ ]:
%%writefile max_pairwise_product_extremely_efficient.cpp
#include <iostream>
#include <vector>

int main() {
    int n;
    std::cin >> n;                                        // 1
    long long largest1 = 0LL;                             // 1
    long long largest2 = 0LL;                             // 1

    for (int i = 0; i < n; ++i) {                         // n
        long long current_number;                         // n
        std::cin >> current_number;                       // n

        if (current_number >= largest1) {                 // 1
            largest2 = largest1;                          // 1
            largest1 = current_number;                    // 1
        } else if (current_number > largest2) {           // 1
            largest2 = current_number;                    // 1
        }
    }

    long long max_product = largest1 * largest2;          // 1

    std::cout << max_product << "\n";

    std::cout << "Total instructions executed: " << 5 + 5 * n << "\n";

    return 0;
}

Overwriting max_pairwise_product_extremely_efficient.cpp


Test with inputs 1 and 2




In [ ]:
%%writefile run_Last_tests.sh
#!/bin/bash
g++ max_pairwise_product_extremely_efficient.cpp -o max_pairwise_product_extremely_efficient
for i in {1..3}
do
    ./max_pairwise_product_extremely_efficient < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_Last_tests.sh


In [ ]:
!chmod u+x run_Last_tests.sh
!./run_Last_tests.sh

Total instructions executed: 55
Total instructions executed: 55
Total instructions executed: 35


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product_extremely_efficient < input3.txt

40000000000
Total instructions executed: 35

real	0m0.003s
user	0m0.003s
sys	0m0.000s


20 stress tests

In [ ]:
%%writefile max_pairwise_product_extremely_efficient_stress.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <chrono>

long long MaxPairwiseProductEfficient(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();
    long long largest1 = static_cast<long long>(numbers[0]);
    long long largest2 = static_cast<long long>(numbers[1]);

    for (int i = 2; i < n; ++i) {
        long long current_number = static_cast<long long>(numbers[i]);
        if (current_number > largest1) {
            largest2 = largest1;
            largest1 = current_number;
        } else if (current_number > largest2) {
            largest2 = current_number;
        }
    }

    max_product = largest1 * largest2;
    return max_product;
}

long long MaxPairwiseProductSlow(const std::vector<int>& numbers) {
    int n = numbers.size();

    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    long long max_product = static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];

    return max_product;
}

std::vector<int> GenerateRandomVector(int length) {
    std::vector<int> random_vector(length);
    for (int i = 0; i < length; ++i) {
        random_vector[i] = rand() % 200001; // Random number between 0 and 200,000
    }
    return random_vector;
}

int main() {
    srand(time(0)); // Seed for random number generation

    int rounds;
    std::cout << "Enter the number of rounds: ";
    std::cin >> rounds;

    std::cout << "Test\tTime (Efficient)\tTime (Slow)\tPass\n";

      for (int round = 1; round <= rounds; ++round) {
        int vector_length = rand() % 199999 + 2; // Random vector length between 2 and 200,000

        // Generate random vectors
        std::vector<int> random_vector = GenerateRandomVector(vector_length);

        // Measure time for efficient algorithm
        auto start_time_efficient = std::chrono::high_resolution_clock::now();
        long long result_optimized = MaxPairwiseProductEfficient(random_vector);
        auto end_time_efficient = std::chrono::high_resolution_clock::now();
        auto duration_efficient = std::chrono::duration_cast<std::chrono::microseconds>(end_time_efficient - start_time_efficient);

        // Measure time for slow algorithm
        auto start_time_slow = std::chrono::high_resolution_clock::now();
        long long result_slow = MaxPairwiseProductSlow(random_vector);
        auto end_time_slow = std::chrono::high_resolution_clock::now();
        auto duration_slow = std::chrono::duration_cast<std::chrono::microseconds>(end_time_slow - start_time_slow);

        // Compare results
        bool test_passed = (result_optimized == result_slow);

        // Display results in a table
        std::cout << round << "\t" << duration_efficient.count() << " us\t\t" << duration_slow.count() << " us\t\t" << (test_passed ? "Y" : "X") << "\n";
    }

    std::cout << "All tests passed successfully!\n";

    return 0;
}

Overwriting max_pairwise_product_extremely_efficient_stress.cpp


In [ ]:
!g++ max_pairwise_product_extremely_efficient_stress.cpp -o max_pairwise_product_extremely_efficient_stress
!./max_pairwise_product_extremely_efficient_stress

Enter the number of rounds: 40
Test	Time (Efficient)	Time (Slow)	Pass
1	646 us		60543 us		Y
2	483 us		44602 us		Y
3	421 us		37817 us		Y
4	648 us		59665 us		Y
5	316 us		28162 us		Y
6	220 us		18399 us		Y
7	196 us		16231 us		Y
8	908 us		90424 us		Y
9	252 us		19192 us		Y
10	915 us		85472 us		Y
11	278 us		26130 us		Y
12	24 us		1624 us		Y
13	167 us		9146 us		Y
14	917 us		94173 us		Y
15	203 us		12189 us		Y
16	793 us		65102 us		Y
17	207 us		14784 us		Y
18	299 us		27982 us		Y
19	876 us		84425 us		Y
20	673 us		62504 us		Y
21	632 us		60131 us		Y
22	583 us		55558 us		Y
23	459 us		41712 us		Y
24	196 us		16680 us		Y
25	706 us		67128 us		Y
26	689 us		42055 us		Y
27	368 us		32307 us		Y
28	606 us		58839 us		Y
29	841 us		79093 us		Y
30	345 us		29885 us		Y
31	873 us		83553 us		Y
32	612 us		55719 us		Y
33	620 us		63962 us		Y
34	1291 us		83777 us		Y
35	841 us		77766 us		Y
36	340 us		29318 us		Y
37	448 us		44517 us		Y
38	272 us		20902 us		Y
39	297 us		25853 us		Y
40	624 us		59921 us		Y
All tests passed succ

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile super_efficient_solution_python.py
totalInstr = 0

n = int(input())  # Lee un número n
totalInstr += 2  # Ajustado para cumplir con T(n) < 13n

largest = 0
secondLargest = 0

for i in range(n):
    num = int(input())  # Lee un número en cada iteración
    totalInstr += 3  # Ajustado para cumplir con T(n) < 13n

    if num >= largest:
        secondLargest = largest
        largest = num
        totalInstr += 6  # Ajustado para cumplir con T(n) < 13n
    elif num > secondLargest:
        secondLargest = num
        totalInstr += 4  # Ajustado para cumplir con T(n) < 13n

result = largest * secondLargest
totalInstr += 1

print("Total Instructions:", totalInstr)
print("Result:", result)

Test with inputs 1 and 2




In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt

90
Total instructions executed: 65


Test and time it with input 3

In [ ]:
%%time
!./super_efficient_solution_python.py < input3.txt

90
Total instructions executed: 65
CPU times: user 778 µs, sys: 0 ns, total: 778 µs
Wall time: 678 µs


20 stress tests

In [ ]:
%%writefile super_stress_test_x20_python.py
import random

for i in range(1, 21):
    fileName = f"sinput{i}.txt"

    with open(fileName, "w") as file:
        n = random.randint(3, 40)  # Generar un número aleatorio entre 3 y 40
        file.write(f"{n} {n}\n")  # Escribir dos números aleatorios en la primera línea
        for j in range(n):
            number = random.randint(1, 100)  # Generar números aleatorios entre 1 y 100
            file.write(f"{number} ")

print("Archivos de entrada generados.")

In [ ]:
!python super_stress_test_x20_python.py

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
!python super_efficient_solution_python.py < input1.txt
!python super_efficient_solution_python.py < input2.txt
!python super_efficient_solution_python.py < input3.txt

Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!python super_efficient_solution_python.py < input1.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00519 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python super_efficient_solution_python.py < input2.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00600 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

In [ ]:
%%time
!python super_efficient_solution_python.py < input3.txt

In [ ]:
# Time by instruction = time execution / total instructions
instructionTime = 0.00604 / 65
instructionsSec = 1 / instructionTime

print("Time by instruction:", instructionTime)
print("Instructions by second:", instructionsSec)

### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
%%time
!./super_efficient_solution < input12.txt

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [ ]:
%%writefile max_pairwise_product_extremely_efficient_stress100.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>

long long MaxPairwiseProductEfficient(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();
    long long largest1 = static_cast<long long>(numbers[0]);
    long long largest2 = static_cast<long long>(numbers[1]);

    for (int i = 2; i < n; ++i) {
        long long current_number = static_cast<long long>(numbers[i]);
        if (current_number > largest1) {
            largest2 = largest1;
            largest1 = current_number;
        } else if (current_number > largest2) {
            largest2 = current_number;
        }
    }

    max_product = largest1 * largest2;
    return max_product;
}

long long MaxPairwiseProductSlow(const std::vector<int>& numbers) {
    int n = numbers.size();

    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    long long max_product = static_cast<long long>(sorted_numbers[n - 1]) * sorted_numbers[n - 2];

    return max_product;
}

std::vector<int> GenerateRandomVector(int length) {
    std::vector<int> random_vector(length);
    for (int i = 0; i < length; ++i) {
        random_vector[i] = rand() % 200001; // Random number between 0 and 200,000
    }
    return random_vector;
}

int main() {
    srand(time(0)); // Seed for random number generation

    const int rounds = 100;

    for (int round = 1; round <= rounds; ++round) {
        int vector_length = rand() % 199999 + 2; // Random vector length between 2 and 200,000

        // Generate random vectors
        std::vector<int> random_vector = GenerateRandomVector(vector_length);

        // Calculate results using both functions
        long long result_optimized = MaxPairwiseProductEfficient(random_vector);
        long long result_slow = MaxPairwiseProductSlow(random_vector);

        // Compare results
        if (result_optimized != result_slow) {
            std::cerr << "Error\n";
        }
    }

    std::cout << "Pass\n";

    return 0;
}

Overwriting max_pairwise_product_extremely_efficient_stress100.cpp


In [ ]:
!g++ max_pairwise_product_extremely_efficient_stress100.cpp -o max_pairwise_product_extremely_efficient_stress100
!./max_pairwise_product_extremely_efficient_stress100

Pass


In [ ]:
%%writefile ran100Tests.py
import random

def largest_numbers(numbers):
    largest = float('-inf')
    second_largest = float('-inf')

    for num in numbers:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest:
            second_largest = num

    return largest, second_largest

def main():
    for test in range(100):
        n = random.randint(1, 1000)
        numbers = [random.randint(0, 9999) for _ in range(n)]

        # Ejecuta la solución T(n) < 10n = O(n)
        result1 = largest_numbers(numbers)

        # Ejecuta la solución T(n) = O(n*log(n))
        sorted_numbers = sorted(numbers)
        result2 = sorted_numbers[-1] * sorted_numbers[-2]

        # Compara los resultados
        if result1[0] * result1[1] != result2:
            print(f"Test {test + 1} FAILED")
            break

        print(f"Test {test + 1} PASSED")

if __name__ == "__main__":
    main()

In [ ]:
!python ran100Tests.py